### PR TITLE
Do not sort qubits by default in pauli_string to fix inconsistencies with pauli_string.gate

### LINKED ISSUE #4270: PauliString class sorts qubits by default, which results in confusing semantics
### Description
`PauliString` represents a collection of pauli's applied on different qubits. Since pauli operators applied to different qubits commute with each other, the class sorts all the qubits by default. Eg:

```python
In [3]: cirq.X(q[1]) * cirq.Y(q[2]) * cirq.Z(q[0])
Out[3]: Z(0)*X(1)*Y(2) # Instead of X(1) * Y(2) * Z(0)
```
Sorting qubits is not common in other operations (eg: `ControlledOperations`, `GateOperations` etc.), and has some confusing semantics as shown below:

### Conversion to/from `DensePauliString` results in different gates
`gate_instance == gate_instance.on().gate()`  should generally be true for all gates, but this is not the case with `DensePauliString`. Eg: 
```python
In [4]: cirq.DensePauliString('XYZ', coefficient=(1+0j)).on(q[1], q[2], q[0]).gate
Out[4]: cirq.DensePauliString('ZXY', coefficient=(1+0j))
```

### `with_qubits` returns a different gate applied on different qubits.  
`with_qubits` function is supposed to return the `same operations applied to different qubits` but the underlying gate also changes with qubit ordering, similar to the same issue as above. Eg:
```python
In [5]: ps = cirq.X(q[0]) * cirq.Y(q[1]) * cirq.Z(q[2])
   ...: ps.gate == ps.with_qubits(q[1], q[2], q[0]).gate
Out[5]: False
``` 

### Need for a `matrix()` protocol. 
Since changing the ordering of the qubits (which happens implicitly) changes the underlying "gate" represented by the operation, the unitary matrix of the operation is not necessarily the same as what one would expect. To get the expected representation, one needs to use `.matrix()`, which makes a case for matrix protocol propsoed in #3099. Eg:

```python
In [6]: m_expected = cirq.linalg.kron(*[cirq.unitary(p) for p in [cirq.X, cirq.Y, cirq.Z]])
    ..: m_unitary = cirq.unitary(cirq.X(q[2]) * cirq.Y(q[1]) * cirq.Z(q[0]))
    ..: m_matrix = (cirq.X(q[2]) * cirq.Y(q[1]) * cirq.Z(q[0])).matrix([q[2], q[1], q[0]])
    ..: np.allclose(m_expected, m_unitary), np.allclose(m_expected, m_matrix)
Out[7]: (False, True)
```

### Proposal
Maybe we should store an explicit `QubitOrder` as well to improve consistency between the operator and underlying gate's representations? 
Update: Note that `QubitOrder` is not serializable and hence it might make more sense to use an `OrderedDict` for `qubit_pauli_map` or an explicit `Iterable[raw_types.Qid]` for storing the `qubit_order`. 

Upd: This is part of organizing gate relationships #3242

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes #4270

Note that the fix relies on the fact that python dicts maintain insertion order from python 3.6+ [[link](https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6)]

### CODE REVIEW

File: cirq-core/cirq/ops/pauli_string_test.py [hunk 1]
```diff
@@ -1350,6 +1350,16 @@ def test_dense():
         _ = p.dense([a, b, d])
 
 
+@pytest.mark.parametrize('qubits', [*itertools.permutations(cirq.LineQubit.range(3))])
+def test_gate_consistent(qubits):
+    g = cirq.DensePauliString('XYZ')
+    print(qubits)
```
Comment (vtomole) [inline cirq-core/cirq/ops/pauli_string_test.py:1356]: Get rid of debug print

Comment (tanujkhattar) [reply cirq-core/cirq/ops/pauli_string_test.py]: done

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
Quoting the python documentation here (https://docs.python.org/3.6/whatsnew/3.6.html#new-dict-implementation):
> The order-preserving aspect of this new implementation is considered an implementation detail and should not be relied upon

So I don't think this as a fix we should be using here.

**Comment by mpharrigan** [comment]
IIRC they changed it in 3.7 to be part of the standard

**Comment by tanujkhattar** [comment]
Yes, it's part of the standard from 3.7+ (see my original link) and hence I think it should be fine to rely on this?

**Review by MichaelBroughton — APPROVED** [review]
Oops, totally missed that. LGTM.

**Pull request merged**